File: examples/dotnet/BatchSend.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Pushfy;

// Send a large audience by splitting it into chunks and calling SendBulkSmsAsync
// once per chunk. Chunking keeps each request within API limits and lets you
// pace the traffic. A per-recipient ext_id keeps every message trackable.
internal static class BatchSendExample
{
    private const int ChunkSize = 500;

    public static async Task Main()
    {
        var token = Environment.GetEnvironmentVariable("PUSHFY_API_TOKEN");
        using var pushfy = new PushfyClient(new PushfyClientOptions { ApiToken = token });

        // Build a demo audience. In production this comes from your database.
        var audience = new List<SmsMessage>();
        for (var i = 0; i < 1200; i++)
        {
            audience.Add(new SmsMessage("5511999999999", $"Batch message #{i}", $"batch-{i}"));
        }

        var chunks = Chunk(audience, ChunkSize);
        Console.WriteLine($"Sending {audience.Count} messages in {chunks.Count} chunks");

        var sent = 0;
        for (var c = 0; c < chunks.Count; c++)
        {
            var batch = chunks[c];
            try
            {
                await pushfy.SendBulkSmsAsync(batch);
                sent += batch.Count;
                Console.WriteLine($"Chunk {c + 1}/{chunks.Count} accepted ({batch.Count} msgs)");
            }
            catch (RateLimitException err)
            {
                var delay = err.RetryAfter.HasValue ? err.RetryAfter.Value * 1000 : 2000;
                Console.Error.WriteLine($"Chunk {c + 1} rate limited; pausing {delay}ms and retrying");
                await Task.Delay(delay);
                await pushfy.SendBulkSmsAsync(batch); // retry once; extIds keep it idempotent
                sent += batch.Count;
            }
            catch (ApiException err)
            {
                Console.Error.WriteLine($"Chunk {c + 1} failed ({err.Status}): {err.Message}");
            }

            // Gentle pacing between chunks.
            await Task.Delay(200);
        }

        Console.WriteLine($"Done. Accepted {sent}/{audience.Count} messages.");
    }

    // Split a list into consecutive sublists of at most `size` elements.
    private static List<List<T>> Chunk<T>(List<T> items, int size)
    {
        var chunks = new List<List<T>>();
        for (var i = 0; i < items.Count; i += size)
        {
            chunks.Add(items.GetRange(i, Math.Min(size, items.Count - i)));
        }
        return chunks;
    }
}

File: examples/java/BatchSend.java
import com.pushfy.ApiException;
import com.pushfy.Pushfy;
import com.pushfy.RateLimitException;

import java.util.ArrayList;
import java.util.List;

/**
 * Send a large audience by splitting it into chunks and calling
 * {@link Pushfy#sendBulkSms(List)} once per chunk.
 *
 * <p>Chunking keeps each request within API limits and lets you pace the
 * traffic. A per-recipient ext_id keeps every message individually trackable.
 */
public class BatchSend {

    private static final int CHUNK_SIZE = 500;

    public static void main(String[] args) throws InterruptedException {
        String token = System.getenv("PUSHFY_API_TOKEN");
        Pushfy pushfy = Pushfy.builder()
                .apiToken(token)
                .build();

        // Build a demo audience. In production this comes from your database.
        List<Pushfy.Sms> audience = new ArrayList<>();
        for (int i = 0; i < 1200; i++) {
            audience.add(new Pushfy.Sms(
                    "5511999999999",
                    "Batch message #" + i,
                    "batch-" + i));
        }

        List<List<Pushfy.Sms>> chunks = chunk(audience, CHUNK_SIZE);
        System.out.println("Sending " + audience.size() + " messages in " + chunks.size() + " chunks");

        int sent = 0;
        for (int c = 0; c < chunks.size(); c++) {
            List<Pushfy.Sms> batch = chunks.get(c);
            try {
                pushfy.sendBulkSms(batch);
                sent += batch.size();
                System.out.println("Chunk " + (c + 1) + "/" + chunks.size()
                        + " accepted (" + batch.size() + " msgs)");
            } catch (RateLimitException err) {
                long delay = err.retryAfter != null ? err.retryAfter * 1000L : 2000L;
                System.err.println("Chunk " + (c + 1) + " rate limited; pausing " + delay + "ms and retrying");
                Thread.sleep(delay);
                pushfy.sendBulkSms(batch); // retry once; extIds keep it idempotent
                sent += batch.size();
            } catch (ApiException err) {
                System.err.println("Chunk " + (c + 1) + " failed (" + err.status
                        + "): " + err.getMessage());
            }

            // Gentle pacing between chunks.
            Thread.sleep(200);
        }

        System.out.println("Done. Accepted " + sent + "/" + audience.size() + " messages.");
    }

    /** Split a list into consecutive sublists of at most {@code size} elements. */
    private static <T> List<List<T>> chunk(List<T> items, int size) {
        List<List<T>> out = new ArrayList<>();
        for (int i = 0; i < items.size(); i += size) {
            out.add(new ArrayList<>(items.subList(i, Math.min(i + size, items.size()))));
        }
        return out;
    }
}
